mlflow/gateway/__init__.py

```diff
@@ -0,0 +1,133 @@
+from mlflow.protos.databricks_pb2 import BAD_REQUEST
+
+import subprocess
+
+server_process: Optional[subprocess.Popen] = None
+gateway_host: Optional[str] = None
+gateway_port: Optional[int] = None
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.DEBUG)
+_stream_handler = logging.StreamHandler()
+_stream_handler.setLevel(logging.DEBUG)
+formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+_stream_handler.setFormatter(formatter)
+_logger.addHandler(_stream_handler)
+_logger.propagate = False
+
+
+def start_server(config_path: str, host, port):
+    """
+    Starts the server with the given configuration, host and port.
+
+    This function sets the global server_process, gateway_host, and gateway_port variables.
+    It starts the server by creating a new subprocess that runs the server script.
+
+    Args:
+        config_path (str): Path to the configuration file.
+        host (str): The hostname to use for the server.
+        port (int): The port number to use for the server.
+
+    Returns:
+        None
+
+    Raises:
+        subprocess.CalledProcessError: If the server subprocess fails to start.
+    """
+    global server_process
+    global gateway_host
+    global gateway_port
+    os.environ[CONF_PATH_ENV_VAR] = config_path
+
+    if server_process is not None:
+        raise MlflowException(
+            f"There is a currently running server instance at pid '{server_process.pid}'. Please "
+            "terminate the running server instance prior to starting a new instance within this "
+            "context.",
+            error_code=BAD_REQUEST,
+        )
+
+    cmd_path = pathlib.Path(__file__).parent.joinpath("gateway_app.py")
+    server_cmd = [sys.executable, str(cmd_path), "--host", host, "--port", str(port)]
+    gateway_host = host
+    gateway_port = port
+    server_process = subprocess.Popen(server_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    time.sleep(SERVER_HOUSEKEEPING_PAUSE)
+
+    if server_process.poll() is not None:
+        stderr_output = server_process.stderr.read().decode()
+        _logger.error("The gateway server has encountered an error and has terminated:")
+        for stderr_line in stderr_output.split("\n"):
+            _logger.error(stderr_line)
+    else:
+        _logger.info("Gateway server is online and is ready to receive requests...")
+    return server_process
+
+
+def _stop_server():
+    """
+    Stops the currently running server.
+
+    This function kills the process of the running server and resets the global server_process
+    variable to None.
+
+    Args:
+        None
+
+    Returns:
+        None
+    """
+    global server_process
+
+    if not server_process:
+        raise MlflowException(
+            "There is no currently running gateway server", error_code=BAD_REQUEST
+        )
+
+    os.kill(server_process.pid, signal.SIGTERM)
```

Comment:
Is it possible for this to throw, e.g. if the server process has already been killed? If so, can we handle that case?

Reply:
excellent point. Adding a process alive check